core/client-internal/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -232,8 +235,28 @@ private void checkAndAdvanceBlockInStream() throws IOException {
           long blockSize = getCurrentBlockSize();
           WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
               mContext.getAluxioBlockStore().getWorkerInfoList(), blockSize);
-          mCurrentCacheStream =
-              mContext.getAluxioBlockStore().getOutStream(currentBlockId, blockSize, address);
+          // Don't cache the block to somewhere that already has it.
+          // TODO(andrew): Filter the workers provided to the location policy to not include
+          // workers which already contain the block. See ALLUXIO-1816.
+          if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
+            WorkerNetAddress readAddress =
+                ((RemoteBlockInStream) mCurrentBlockInStream).getWorkerNetAddress();
+            // Try to avoid an RPC.
+            if (readAddress.equals(address)) {
+              mShouldCacheCurrentBlock = false;
+            } else {
+              BlockInfo blockInfo = mContext.getAluxioBlockStore().getInfo(currentBlockId);
+              for (BlockLocation location : blockInfo.getLocations()) {
+                if (address.equals(location.getWorkerAddress())) {
+                  mShouldCacheCurrentBlock = false;
```

Comment:
It will be reset whenever `updateBlockInStream` is called, which happens whenever we switch blocks.